class-two-factor-core.php

```diff
@@ -1230,6 +1289,144 @@ public static function login_form_validate_2fa() {
 		exit;
 	}
 
+	/**
+	 * Display the "Revalidate Two Factor" page.
+	 *
+	 * @since 0.8.0
+	 */
+	public static function login_form_revalidate_2fa() {
+		$provider    = ! empty( $_REQUEST['provider'] )    ? sanitize_text_field( wp_unslash( $_REQUEST['provider'] ) ) : false;
+		$redirect_to = ! empty( $_REQUEST['redirect_to'] ) ? wp_unslash( $_REQUEST['redirect_to'] )                     : admin_url();
+
+		if ( ! is_user_logged_in() ) {
+			wp_safe_redirect( home_url() );
+			exit;
+		}
+
+		$user            = wp_get_current_user();
+		$session_token   = wp_get_session_token();
+		$session_manager = WP_Session_Tokens::get_instance( $user->ID );
+		$session         = $session_manager->get( $session_token );
+		$providers       = self::get_available_providers_for_user( $user );
+
+		// Default to the currently logged in provider.
+		if ( ! $provider && ! empty( $session['two-factor-provider'] )	) {
+			$provider = $session['two-factor-provider'];
+		}
+
+		if ( $provider && isset( $providers[ $provider ] ) ) {
+			$provider = $providers[ $provider ];
+		} else {
+			$provider = self::get_primary_provider_for_user( $user->ID );
+		}
+
+		if ( ! $provider  ) {
+			wp_die( esc_html__( 'Cheatin&#8217; uh?', 'two-factor' ), 403 );
+		}
+
+		// Run the provider processing.
+		$result = self::process_provider( $provider, $user, 'revalidate' );
+
+		if ( true !== $result ) {
+			$error = '';
+			if ( is_wp_error( $result ) ) {
+				do_action( 'wp_login_failed', $user->user_login, $result );
+
+				$error = $result->get_error_message();
+			}
+
+			self::login_html( $user, '', $redirect_to, $error, $provider, 'revalidate_2fa' );
+			exit;
+		}
+
+		$session['two-factor-provider'] = get_class( $provider );
+		$session['two-factor-login']    = time();
+
+		$session_manager->update( $session_token, $session );
+
+		// Must be global because that's how login_header() uses it.
+		global $interim_login;
+		$interim_login = isset( $_REQUEST['interim-login'] ); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited,WordPress.Security.NonceVerification.Recommended
+
+		if ( $interim_login ) {
+			$message       = '<p class="message">' . __( 'You have revalidated successfully.', 'two-factor' ) . '</p>';
+			$interim_login = 'success'; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+			login_header( '', $message );
+			?>
+			</div>
+			<?php
+			/** This action is documented in wp-login.php */
+			do_action( 'login_footer' );
+			?>
+			</body></html>
+			<?php
+			exit;
+		}
```

Comment:
I'm not sure if this interim-login functionality is useful to keep in here or not, as while it provides a better in-line UI for revalidating, it doesn't provide any way of reloading the parent page / changing the state of the parent pages form.